kythe/docs/rfc/2967.md

```diff
@@ -0,0 +1,64 @@
+## Build Configuration in the Kythe Graph
+
+[TOC]
+
+## Objective {#objective}
+
+Incorporate build configuration information into the Kythe graph. Somewhat relatedly, provide canonical guidelines on how to identify build-time generated artifacts.
+
+
+## Background {#background}
+
+Currently, the Kythe graph largely assumes that it is built for a single configuration, at a consistent revision for a particular corpus.  Historically, this has worked well.  But as time marches onward and the diversity of platforms built from a single corpus has grown it has left a widening gap in the information available to users with only partial cross references for some files (due to conditional compilation) or other files entirely omitted due to build-time selection.  In order to work around this lack, some third-party corpora (most notably Chromium) have begun to use the "root" VName field to encode this information.  Similarly, the extant "prodkernel" corpus has need of this, but currently makes do without. We reasonably anticipate that other soon-to-be-corpora such as Android and Tensorflow would encounter similar limitations.
+
+
+## Overview {#overview}
+
+Rather than introduce any new VName fields, nodes or edges build configuration can be adequately represented via a signature-relevant fact on anchor nodes and some ancillary clarifications to the guidelines for root and corpus.  This is predicated on:
+
+
+1.  Non-generated source files exist independent of build configuration.
+1.  Nodes must already incorporate relevant information into their signature.  This includes any elements related to build-configuration which materially alters the represented entity. This is currently already done by non-anchor (aka "semantic") nodes, but build configuration has not been considered relevant for anchors.
+1.  Certain languages will necessarily emit different anchors under different build configurations (C/C++ are particularly subject to this).
```

Comment:
I don't get this example. C++ emits different anchors because the compiler compiles the headers differently, and/or stores output in different directories. Is there some information we are currently adding to anchors _without_ the addition of new build-config tags, that we would be able to get rid of by including it?

Reply:
My question was to relevancy: This bullet point offers itself as partial explanation for why a new fact is required for anchors. But I don't see the connection between "certain languages will need to emit different anchors under different build configurations" when the example language—C++—already _does_ emit different anchors (with or without the distinction). So my question was that if it's _not_ about distinguishability, then I don't know why this point supports the design.

Reply:
It emits some different anchors and some "identical" anchors.  There is no way of distinguishing the two at present and (importantly) no way for a user to do so.  It is about distinguishability, but that was not the question (which was about "things we can remove" due to this change).